include/node.hpp

```diff
@@ -733,8 +820,28 @@ class node : protected std::tuple<Arguments...> {
     work(std::size_t requested_work = std::numeric_limits<std::size_t>::max()) noexcept {
         constexpr bool is_blocking = node_template_parameters::template contains<BlockingIO>;
         if constexpr (is_blocking) {
-            return work_internal(requested_work);
-            // static_assert(fair::meta::always_false<derived_t>, "not yet implemented");
+            std::atomic_store_explicit(&ioRequestedWork, requested_work, std::memory_order_release);
+            if (bool expectedThreadState = false; ioThreadShallRun.compare_exchange_strong(expectedThreadState, true, std::memory_order_acq_rel)) {
+                ioThreadPool->execute([this]() {
+                    ioThreadRunning.fetch_add(1);
+                    while (ioThreadShallRun) {
+                        auto [work_requested, work_done, last_status] = work_internal(ioRequestedWork.load(std::memory_order_relaxed));
+                        ioWorkDone.increment(work_requested, work_done);
+                        ioLastWorkStatus.exchange(last_status, std::memory_order_relaxed);
+
+                        std::ignore = progress->incrementAndGet();
+                        progress->notify_all();
+                        if (last_status == work_return_status_t::DONE) {
+                            ioThreadShallRun.exchange(false, std::memory_order_relaxed);
+                        }
+                    }
+                    ioThreadRunning.fetch_sub(1);
+                    ioThreadRunning.notify_all();
+                });
+            }
+            const work_return_status_t lastStatus                 = ioLastWorkStatus.exchange(work_return_status_t::OK, std::memory_order_relaxed);
+            const auto &[accumulatedRequestedWork, performedWork] = ioWorkDone.getAndReset();
+            return { accumulatedRequestedWork, performedWork, performedWork > 0 ? work_return_status_t::OK : lastStatus };
```

Comment:
I localize now the problem with `emsripten`
For the moment I could fix the problem by changing this lne to 
`return { accumulatedRequestedWork, performedWork, lastStatus };`.
But I need some time to understand why `performedWork` is always > 0.

Reply:
One solution is to implement `available_samples` method for `ClockSource` class:
```cpp
    constexpr std::make_signed_t<std::size_t>
    available_samples(const ClockSource & /*self*/) noexcept {
        const auto ret = static_cast<std::make_signed_t<std::size_t>>(n_samples_max - n_samples_produced);
        return ret > 0 ? ret : -1; // '-1' -> DONE, produced enough samples
    }
```
With this change ll test passed on my computer.

Reply:
Also one needs to be accurate with a `chunk_size` and add some additional checks:
```cpp
    constexpr std::make_signed_t<std::size_t>
    available_samples(const ClockSource & /*self*/) noexcept {
        const auto ret = static_cast<std::make_signed_t<std::size_t>>(n_samples_max - n_samples_produced);
        if (ret < chunk_size) return -1;
        return ret > 0 ? ret : -1; // '-1' -> DONE, produced enough samples
    }
```

and in `process_bulk` function:
```cpp
        if (writableSamples < chunk_size && n_samples_max - n_samples_produced < chunk_size) {
            output.publish(0_UZ);
            return work_return_status_t::DONE;
        }
```

But with these checks one can produce less samples and tests fail.

Reply:
Here is a branch https://github.com/fair-acc/graph-prototype/tree/deadlock_emscripten

Reply:
@drslebedev we want to get rid of the `available_samples(..)` functions for sources. I fond a workaround and pushed again. Should work now. 

Your past decimator/interpolator+stride PR seems to have some regression w.r.t. that the number of samples that are processed as  a batch by `process_bulk(...)` and `process_one(...)` isn't limited to the min of available input samples and the next sample with an attached tag. I commented the corresponding test. Maybe you could investigate and fix this after this PR is merged.